### PR TITLE
Feature/add function to handle error in action

### PR DESCRIPTION
**Pull request by ouaibsky**
Please note:
* This is first version,
* API is compatible with previous version.
* Files where tab indented but guidelines require spaces, so modified files have been updated with spaces
* Add basic test
* Introduce new AbstractBasicTransition class to manage common code between InitalTransition and AbstractTransition, maybe there is better name ...
* Basic unit test to check that error action is called when provided by *Configurer
* make some cleaning in InitialTransition constructors

Christophe

### CODE REVIEW

File: spring-statemachine-core/src/main/java/org/springframework/statemachine/action/Actions.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+     *
+     * @param <S>
+     * @param <E>
+     * @return an empty (Noop) Action.
+     */
+    public static <S, E> Action<S, E> emptyAction() {
+        return new Action<S, E>() {
+            @Override
+            public void execute(final StateContext<S, E> context) {
+                // Nothing to do;
```
Comment (ouaibsky) [inline spring-statemachine-core/src/main/java/org/springframework/statemachine/action/Actions.java]: maybe a debug log statement ?

File: spring-statemachine-core/src/main/java/org/springframework/statemachine/action/Actions.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package org.springframework.statemachine.action;
```
Comment (jvalkeal) [inline spring-statemachine-core/src/main/java/org/springframework/statemachine/action/Actions.java]: New files should have license header.

File: spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractBasicTransition.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
+package org.springframework.statemachine.transition;
```
Comment (jvalkeal) [inline spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractBasicTransition.java]: New files should have license header.

File: spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractTransition.java [hunk 4]
```diff
@@ -35,83 +36,68 @@
...
 
-	private final State<S,E> source;
-
-	private final State<S,E> target;
+	private final State<S, E> source;
 
-	private final Collection<Action<S, E>> actions;
-
-	private final TransitionKind kind;
+    private final TransitionKind kind;
```
Comment (jvalkeal) [inline spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractTransition.java]: spaces to tabs.

File: spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractTransition.java [hunk 5]
```diff
@@ -35,83 +36,68 @@
...
 		Assert.notNull(kind, "Transition type must be set");
 		this.source = source;
-		this.target = target;
-		this.actions = actions;
 		this.kind = kind;
 		this.guard = guard;
 		this.trigger = trigger;
 		this.securityRule = securityRule;
-	}
+    }
```
Comment (jvalkeal) [inline spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractTransition.java]: spaces to tabs

**Comment by jvalkeal** [comment]
From a plain compile point of view, should fix javadoc warnings:
```
:api/home/jvalkealahti/repos/tmp/ssm/spring-statemachine/spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/AbstractBasicTransition.java:41: warning: no @param for errorAction
	public AbstractBasicTransition(State<S, E> target, Collection<Action<S, E>> actions,
	       ^
/home/jvalkealahti/repos/tmp/ssm/spring-statemachine/spring-statemachine-core/src/main/java/org/springframework/statemachine/transition/Transition.java:70: warning: no description for @return
	 * @return
	   ^
/home/jvalkealahti/repos/tmp/ssm/spring-statemachine/spring-statemachine-core/src/main/java/org/springframework/statemachine/config/builders/StateMachineTransitionBuilder.java:148: warning: no @param for errorAction
	public void addTransition(S source, S target, S state, E event, Long period, Integer count, Collection<Action<S, E>> actions,
	            ^
/home/jvalkealahti/repos/tmp/ssm/spring-statemachine/spring-statemachine-core/src/main/java/org/springframework/statemachine/action/Actions.java:16: warning: no description for @param
     * @param <S>
       ^
/home/jvalkealahti/repos/tmp/ssm/spring-statemachine/spring-statemachine-core/src/main/java/org/springframework/statemachine/action/Actions.java:17: warning: no description for @param
     * @param <E>
       ^
/home/jvalkealahti/repos/tmp/ssm/spring-statemachine/spring-statemachine-core/src/main/java/org/springframework/statemachine/config/model/TransitionData.java:100: warning: no @param for errorAction
	public TransitionData(S source, S target, S state, E event, Long period, Integer count, Collection<Action<S, E>> actions,
	       ^

6 warnings
```

and remove willdcard imports.

**Comment by jvalkeal** [comment]
You can probably hold making any fixes to this PR until I've actually poked around it a bit more on a code level.

**Comment by ouaibsky** [comment]
about indenting is it space or tabs ? (because in guidelines it's written spaces)

**Comment by jvalkeal** [comment]
What guidelines you're referring to? If you mean docs, then yes all tabs are replaced by spaces as tabs are bad in html formatting. We should probably be more clear on this, but generally all spring umbrella projects follow how spring framework does things. You have no idea how many internal email threads we have had about code formatting, as spring devs can make a fight out from anything :)

**Comment by jvalkeal** [comment]
Ok, some comments:

## AbstractTransition
When we create an abstract class implementing an interface then either it has methods implemented or defined as abstract function. This is what we had in
```
public abstract class AbstractTransition<S, E> implements Transition<S, E> {
```
Making `AbstractBasicTransition` as parent of `AbstractTransition` is redundant as everything could go to existing class. 

## handling error
You only added action error handling to action executed during a transition. As we also have actions executed during state entry/exit and in a state itself(doActivity behaviour). Adding this same try/catch logic in various places feels a bit dump copy/pasting of that code. What about if we use a concept from [ExceptionCatchingAction](https://github.com/spring-cloud/spring-cloud-deployer-yarn/blob/master/spring-cloud-deployer-yarn/src/main/java/org/springframework/cloud/deployer/spi/yarn/AbstractDeployerStateMachine.java#L298) which can wrap an action and handle try/catch there and call errorAction if defined. Then if user passes errorAction to configuration we could automatically wrap original action with it and all error logic would be in a same place.

And btw, in your `executeAllActions()` when you execute `errorAction.execute(context)`, that should also be within a try/catch again.

**Comment by ouaibsky** [comment]
* For enter/exit I'll do it into another commit / PR. I prefer to have small piece of code to make review as simple and fast as possible.
* about guidelines I'm referring what is into statemachines README: https://github.com/spring-projects/spring-statemachine/blob/master/CONTRIBUTING.md.

**Comment by jvalkeal** [comment]
Yeah, it's ok to make smaller PR's(just made a not). Maybe we can log error in a different place and anyway when we move from 1.0.x to 1.1.x we can make these kind of changes. With point releases we should not change any behaviour if not needed.

I'm currently on PTO(hiking in wales) but I'll try to handle these as much as I can.

**Comment by jvalkeal** [comment]
I think this look good for now and we can start to implement other mentioned issues in other PR's. Merging.

**Comment by jvalkeal** [comment]
Squashed/merged per https://github.com/spring-projects/spring-statemachine/commit/4c34e46ccfa9db7ef06fa76c592f927909ca9b51

Thanks!

**Comment by ouaibsky** [comment]
ok Thx, I'll continue with enter/exit Action for State.
Christophe

**Pull request closed without merging**